docs/modules/customize/pages/logic.adoc

```diff
@@ -0,0 +1,114 @@
+* In the copied file, add/edit a validation for maximum allowable value:
+[source,ruby]
+----
+#...
+validates :budget_amount, numericality: { less_than_or_equal_to: :maximum_budget }
+#...
+private
+#...
+  def maximum_budget
+    (budget[:total_budget] * 0.5).round
+  end
+----
+* Save the file and restart the server from the console (just to be sure), and check if the changes has taken effect by trying to submit the form.
+* Try to create a project with the budget of €5,001 for a budget with the total amount of €10,000 (e.g. "My budget").
+* You should see an error under the project budget field after submitting the form if everything went as expected.
+
+[.maximum-budget-rules]
+.Error messages indicating the exceeding the maximum budget rule.
+
+image::error_max_budget.png[Error for reaching maximum budget]
+
+== 2. Override logic with a concern
+
+Concerns allow us to add functionality into existing classes so that the including class can use the added functionality. This is a pattern provided by the Rails framework and you can learn more about it from the https://api.rubyonrails.org/classes/ActiveSupport/Concern.html[Rails API documentation]
+Simply put, a concern allows developers to change only the parts of the original code that need to be changed. This might be a block of a code we need to change, or add/edit an existing method, etc.
+
+The benefit of using concerns is that when we update Decidim gems to new versions, we do not have to necessarily go through all our customizations and we can just focus on those parts that we really want to change. This is slightly harder to do than the first example but in the long run it will pay you back as you will save a lot of time when updating Decidim.
+
+Concerns should be placed into the `concerns` folder within the top-level folder where you want to apply these concerns. In this case, we are adding a concern for the forms classes, so we should place the concerns related to forms to `app/forms/concerns`. If you do not have this folder available, create it first `mkdir -p app/forms/concerns`.
+
+Now, let's move our modifications from the first copy-paste example into a concern.
+
+Create the file `app/forms/concerns/admin_project_form_extensions.rb`, and make the desired changes in that file as follows:
+
+[source,ruby]
+----
+# frozen_string_literal: true
+module AdminProjectFormExtensions
+  extend ActiveSupport::Concern
+    included do
+      validates :budget_amount, numericality: { less_than_or_equal_to: :maximum_budget }
+    end
+  private
+    def maximum_budget
+      (budget[:total_budget] * 0.5).round
+    end
+end
+----
+After the concern is created, we need to apply it to the correct class. We can do this in the `to_prepare` hook in `config/application.rb`:
+
+[source,ruby]
+----
+config.to_prepare do
+  Decidim::Budgets::Admin::ProjectForm.include(Decidim::Budgets::Admin::ProjectForm.include(AdminProjectFormExtensions)
+end
+----
+
```

Comment:
Let's also add here another note if you want to change anything in the controllers:

```suggestion
This applies to almost all classes you want to change in Decidim but controllers are a bit special case due to order of how things are loaded in the Decidim boot process. If you want to add any changes to controllers, you need to wrap the `config.to_prepare` block within an initializer that is run at the correct phase of the boot process as follows:

[source,ruby]
----
initializer "customizations", after: "decidim.action_controller" do
  config.to_prepare do
    # Add the controller customizations here
  end
end
----

```